Validate string arguments to transaction

dariorussi opened the pull request:
### Description
Here we are again...
this solution has (or should have) some significant benefits over the previous one. 
Please review and find holes!

The most important change is that now we only walk and verify string arguments. Anything else is simply skipped. That should avoid the problem with the old implementation and save us from walking and spending time over arguments that do not matter.
We also check the overall size as we go and if we get over `static MAX_STRING_ARGS_SIZE: usize = 64 * 1024;` we stop and report a bad argument. A note about this limit:
- if the arguments are in fact correct, the overall transaction size check is before any function in args verification (the file changed) is invoked and the transaction fails for being too big. I believe that is 32K so the bigger value does not matter.
- if the argument is bad and a fake length is reported, the limit is probably more than we care but at the same time we never walk more than transaction max size character to check for valid utf8 (as in the previous point). We have tests that verify that. Overall that seems reasonable.

I thought we allowed for generic/templetized entry functions but that does not appear to be the case. We have a test for that (as I thought it was legal) which checks for failure, I am not sure if that is more noise than value. Please let me know and happy to remove it.

Overall we are giving some free computation to people. 32K or so of character checks are super cheap so I do not believe that is an issue (thoughts?) and we should charge for transaction size anyway. 
Bottom line we walk at most a valid transaction size characters and that should not be an issue (thoughts?).

There are a couple of open questions about an `expect` and a failure over an "impossible" condition. 
I would love some help/comments on that.
Should we change the `expect` to failures? that would simply avoid a `panic!` though again it should not be possible,
I am returning an error over another impossible condition, that needs logging at least (can someone tell me how we log warning or errors?), but people may also be fine with a `panic!`.
Suggestions welcome!

### Test Plan
Added a bunch of tests to verify correctness of solution.
I wanted to add some unit tests that do not involve running the VM as that is very expensive. Also that would be a good way to thoroughly check we are doing the right thing. A proptest style test for that would in fact be ideal.
But I did not have time to do that yet and its getting late so I am putting this here anyway to collect feedback

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/aptos-labs/aptos-core/4577)
<!-- Reviewable:end -->

dariorussi commented on aptos-move/aptos-vm/src/transaction_arg_validation.rs:
```diff
@@ -83,22 +101,133 @@ pub(crate) fn validate_combine_signer_and_txn_args<S: MoveResolverExt>(
+            // performed in `is_valid_txn_arg`
+            let st = session
+                .get_struct_type(*idx)
+                .expect("unreachable, type must exist");
```
this and below, should we return a failure instead?
this is the impossible condition I mentioned i the main comment, but....

dariorussi commented on aptos-move/aptos-vm/src/transaction_arg_validation.rs:
```diff
@@ -83,22 +101,133 @@ pub(crate) fn validate_combine_signer_and_txn_args<S: MoveResolverExt>(
+        // In other words this should never happen but we are conservatively
+        // reporting an error.
+        // TODO: log something
+        _ => return Err(VMStatus::Error(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT)),
```
here we are returning an error over a condition that should be impossible.
What is a proper log to add? People happy with the error?

zekun000 replied:
I'm fine with unreachable

dariorussi commented:
please add reviewers as you see fit...

dariorussi commented on aptos-move/aptos-vm/src/transaction_arg_validation.rs:
```diff
@@ -56,10 +69,15 @@ pub(crate) fn validate_combine_signer_and_txn_args<S: MoveResolverExt>(
     // validate all non_signer params
-    for ty in func.parameters[signer_param_cnt..].iter() {
-        if !is_valid_txn_arg(session, ty) {
+    let mut needs_validation = vec![];
```
should we change this to a `Set`?
I run under the assumption a worst case scenario here cannot be that big anyway (this is a function of the number of arguments to an entry), but maybe we should do it for "safety"?

wrwg replied:
I think it is fine

dariorussi commented on aptos-move/aptos-vm/src/transaction_arg_validation.rs:
```diff
@@ -83,22 +101,133 @@ pub(crate) fn validate_combine_signer_and_txn_args<S: MoveResolverExt>(
-fn is_valid_txn_arg<S: MoveResolverExt>(session: &SessionExt<S>, typ: &Type) -> bool {
+// Return whether the argument is valid/allowed and whether it needs validation.
+// Validation is only needed for String arguments at the moment and vectors of them.
+fn is_valid_txn_arg<S: MoveResolverExt>(session: &SessionExt<S>, typ: &Type) -> (bool, bool) {
```
this and all recursive related functions here are assumed to be "bounded" by the nested types in the signature allowed. 
I am not sure if that is a safe assumption or if we should put a limit to the recursion here

wrwg replied:
Lets remember this for another PR. In generally we wanted to take a close look at recursion and bounds, but not know.

wrwg reviewed (approved):
This looks great.

I wonder whether we want to ensure this gets into devnet (build tomorrow) so we can battle test. @davidiw

zekun000 commented on aptos-move/aptos-vm/src/transaction_arg_validation.rs:
```diff
@@ -83,22 +101,133 @@ pub(crate) fn validate_combine_signer_and_txn_args<S: MoveResolverExt>(
+        // In other words this should never happen but we are conservatively
+        // reporting an error.
+        // TODO: log something
+        _ => return Err(VMStatus::Error(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT)),
```
nit: I'd list all the rest variants here instead of using `_`

zekun000 commented on aptos-move/aptos-vm/src/transaction_arg_validation.rs:
```diff
@@ -83,22 +101,133 @@ pub(crate) fn validate_combine_signer_and_txn_args<S: MoveResolverExt>(
+        Vector(inner) => {
+            // get the vector length and iterate over each element
+            let mut len = get_len(cursor)?;
+            *current_size_for_strings += len;
```
this is not the string size but the number of strings? is this intended?

dariorussi replied:
hmmm now you are making me think about all of this...

in general the name may have not been the best but it was intended to cover everything string related. So in a vector of vectors of strings it would not just include the string size but also the vector size, that was the intent.
But now I am realizing this is not the most accurate computation and maybe it's all good given the constant and what we do but let's think about it a bit.
If I get a vector of size 100 and 0 size string this would consider 100, which is good, but it's also not considering the space for the string length. So maybe adding 1 to the string length could be better, that would cover strings of zise 0 even when not in a vector. But the vector of vectors has to be taken into account. 
Moreover by just using the length we are not really computing the proper size because the number of bytes used for bigger numbers is more than 1. 
So on a vector of vectors, if each nested vector is of size > 2^7, then each size would take more than 1 byte and if the outer was length 100, the overall usage would be at least 200.
But I also wanted to be conservative and if the overall size is bigger than our constant 64K I wanted to stop no matter what.

Overall I think this may be just fine in the sense that we have a couple fo situations:
- the size of the payload is bigger then max size in which case we fail and never enter this code
- the payload is bogus and sizes are fake. In that case we stop under one of the 2 cases: 1- we reach the end of the arguments payload 2- we do not even look because we determined that the overall length is bigger than 64K. The latter seems just a early exit at this point.

My current feeling is that the length check could be easily moved out and rely on the end of the payload as a termination. In other words I do not see how we can create problems even with fake lengths.
Overall though I also think this is fine and I do not see how it would make a valid case invalid. That is, the early termination is just a minor optimization.

Happy to remove it. Or leave things as they are.

As for the variable name it seems ok to me but also happy to change it

zekun000 replied:
personally I'd like to have explicit limit on both number of items, and individual item limit. for example, we allow at most 64 strings/vector elements and each string can have at most 255 length or something like that

dariorussi replied:
I really do not like that - lol
why would you think that is a fine story? 
It seems to limit usage with no particular advantage. If I remember correctly we already know of input strings that are of a significant size, I have hard time thinking how to explain people the limitation.
Also would that apply to strings only? I can pass any size bytearray but only a certain size string?
Don't know, I have hard time to make sense of the limitation....
thoughts?

zekun000 replied:
I agree limiting only strings looks weird, ideally we should have a well-defined arguments limit, but probably not for this PR

davidiw replied:
I think the length limit should be related to the allowable txn size limit. We currently support a 64K size limit. We should note this accordingly. The name ought to be current_evaluated_length or something

davidiw replied:
note that length does not indicate the length size
perhaps we should be focused on the change to the cursor?

dariorussi commented:
addressed feedback about `unreachable` and explicit `match` arms.
Please review comment about length and let me know your thoughts

wrwg commented:
Would be great if we can have this in upcoming devnet cut. @sherry-x @davidiw

davidiw reviewed (dismissed):
This needs to get in, but we should do it through gatekeepers review. @zekun000 let's make sure one of us are comfortable with a second approval beforehand.

davidiw commented on aptos-move/aptos-vm/src/transaction_arg_validation.rs:
```diff
@@ -83,22 +101,133 @@ pub(crate) fn validate_combine_signer_and_txn_args<S: MoveResolverExt>(
+        // this is unreachable given the check in `is_valid_txn_arg` and the
+        // fact we collect all arguments that involve strings and we validate
+        // them and them only
+        Bool | U8 | U64 | U128 | Address | Signer | Reference(_) | MutableReference(_)
```
I see, so we're still not evaluating the cursor... so we could have actually ended earlier if we evaluate the cursor length here too. The primary issue was not the size of the actual size of the txn, it was that we didn't properly mock a deserialization.

davidiw replied:
We should at least take into consideration size_of for the curent_size_for_strings -- most of these are many bytes big.

davidiw reviewed (commented):
I'd prefer to see something that just mirrors deserialization with a maximum number of bytes. The bug we found would have been found faster by just validating there was sufficient data in the cursor.

dariorussi commented:
updated to a revisited version that I hope addresses some of the latest concerns (particularly as from @davidiw).
I removed the max size because it is a huge redherring and it is not keeping us focus on the issue. Having said so - and I will explain later - @davidiw has a valid argument that I hope is addressed now.

Please bear with me as I go through the logic of this diff and the flow as I see it.

First off, this code is intended to check only strings. So with the latests changes (latest as in few days now) we only check arguments that contain a string, that is `String` and `vector<vector<...<vector<String>..>>`, any other argument is simply ignored now. I think that is the right thing to do and it would otherwise be a duplication of work we do later.
Please notice that the above is allowed and possible because we receive the arguments as a `Vec<Vec<u8>>` where the outer vector is the size of the arguments. So if we have 3 arguments the length of the outer vector would be 3 and each inner vector would be the serialized form of the specific argument. That is vital to understand because it is what allows us to check only arguments with `String`. Again all of that seems very good to me. If we have issues with that, this diff is not addressing any of those.

The code here is exercised/run only if the size of the transaction payload is correct and legal (not over the max size we allow), and this code runs after the prologue has run already. In that respect the issues we need to cover are with a bogus or adversarial input.

Before we go there (bogus/adversarial payload), an important note on valid inputs.
The most "aggressive" vector of attack IMO is to pass a `vector<String>` (or a more nested one) with a lot of strings and the last one being an invalid utf8 string. As I'll hopefully explain that is more annoying and problematic than any bogus payload one can create. We have tests for the scenario I just describe and I measured the time taken (albeit with `Instant` and `--test-threads=1` for what that is worth) and we take up to 5ms to check correctness. That seems a lifetime to me and considering that is "free" time we give to users, a correct input with a bad last string would be the worst that can happen. In that case the user will pay for the transaction size and we would have given them free computation in the form of few milliseconds (again a lifetime I think). Not sure if we should charge accordingly, if we do charge properly already, or if we should absorb the cost for now (as we do for other operation on the input that we perform already - all non gas related computation).

On bogus inputs. 
I think the code was doing already what @davidiw was suggesting (I think) in the sense that the cursor would fail if it was asked to go over the buffer. In other words both `get_len` (line 211 in `transaction_arg_validation.rs`) and `read_exact` (line 190) would fail because they would go over the size of the argument buffer. And again given that the payload must have been correct we would at most read a "legal" size buffer, which seems inevitable.
There was however a problem that I did not spot (and what @davidiw hinted to, I think, when he says "mimicking serialization") and that is with line 188 in `transaction_arg_validation.rs`: `let mut s = vec![0u8; len];` which allocates the buffer to read the argument into (`read_exact` input, I believe serialization would do that too? maybe check before as we do now?). With a bogus transaction payload and with no checks for that, even though the read would fail, the allocation would still be there, possibly causing massive problems. That allocation is now protected by line 177
```
            let current_pos = cursor.position() as usize;
            let len = get_len(cursor)?;
            match current_pos.checked_add(len) {
                Some(size) => {
                    if size > arg_len {
                        return Err(VMStatus::Error(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT));
                    }
                }
                None => return Err(VMStatus::Error(StatusCode::FAILED_TO_DESERIALIZE_ARGUMENT)),
            }
```
in other words we check that there are enough characters in the input to satisfy the `read_exact` and if that is not the case fail without allocating anything (a bunch of tests verify that).
What about the vector loop?
```
            let mut len = get_len(cursor)?;
            while len > 0 {
                validate_arg(session, inner, cursor, arg_len)?;
                len -= 1;
            }
```
there is no check there and what happens with a bogus input that gets a length of `usize::MAX`? should we add the check we do for strings here too?
My take is that we would at most loop over the size of the payload (remember `get_len` both in vectors and strings does increase the position in the buffer/cursor even with `len` of `0`) and that is not worst (in fact slightly better) that the valid payload I talked about (a vector with lots of strings with last one invalid). Given that, I did not feel adding the check would give us much, but happy to add it there too if people feel more comfortable with that.

Finally a note on recursion. As commented before recursion is not controlled by the arguments but rather the signature of the function. So we recurse in as much as the type in the signature allows (`vector<vector<vector<....>>>`). I assumed we do not need to check for that (as in checked elsewhere) and if that is an issue we have the problem here and probably elsewhere. The feedback I got was to let that go for now, not sure what the overall feeling is.

dariorussi commented:
on the time spent I realize that is just with `cargo build` not sure if that builds a release or not, I will look into it but if someone knows please let me know

davidiw reviewed (commented):
good for me

arg_len is an optimization, it looks right, but ...

dariorussi commented:
> arg_len is an optimization, it looks right, but ...

it's not just an optimization is pretty vital for this to work right?
what bothers you, please share...

davidiw commented:
> > arg_len is an optimization, it looks right, but ...
> 
> it's not just an optimization is pretty vital for this to work right? what bothers you, please share...

Not strictly necessary

dariorussi commented:
I did a small cleanup, moved code to a more proper place (basically 5 lines are moved few lines down close to their usage.
I swapped a get position function after we get length so it's more correct (it did not cause any issue but it's more proper the way it is now).
And I added tests both good and bad for big vector of vector.
Speed in release mode is better but it's still very concerning how slow we can get with complex arguments in input. 
I think it's something to talk about and do something about.
Maybe @zekun000 idea to cut down on what we allow for arguments is not that bad....

dariorussi commented:
> > > arg_len is an optimization, it looks right, but ...
> > 
> > 
> > it's not just an optimization is pretty vital for this to work right? what bothers you, please share...
> 
> Not strictly necessary

well not really given that if it's not there we would allocate a vector of whatever size a malicious user wants and crash...
so I'd say vital... thoughts?

zekun000 commented on aptos-move/aptos-vm/src/transaction_arg_validation.rs:
```diff
@@ -83,22 +96,134 @@ pub(crate) fn validate_combine_signer_and_txn_args<S: MoveResolverExt>(
+}
+
+// Validate a single arg. A Cursor is used to walk the serialized arg manually and correctly.
+// Only Strings and nested vector of them are valiadated.
```
"validated"

dariorussi commented:
fixed comment.
I keep getting sdk files changed on my machine and sdk problem with the diff, I have no idea what's going on but we'll see what happens....

davidiw commented:
Please update ASAP

dariorussi commented:
hopefully this resolves all the problems with this and that I got on my machine after a rebase

jahid19971 commented:
#Martian #Aptos #Wallet #Mainnet #Live #Launch
